src/main/java/duke/tasks/TaskList.java

```diff
@@ -0,0 +1,306 @@
+        this.list.get(index).markAsDone();
+        this.exportData();
+        return String.format("Nice! I've marked this task as done:%n%s", this.list.get(index).toString());
+    }
+
+    /**
+     * Marks a task as undone.
+     *
+     * @param num The number of the task to be marked as undone.
+     * @return String message
+     * @throws DukeIllegalArgumentException If the task number is out of range of the list.
+     */
+    public String unmark(int num) throws DukeIllegalArgumentException {
+        int index = num - 1;
+        if (index < 0 || index >= this.list.size()) {
+            throw new DukeIllegalArgumentException(ERROR_MESSAGE_INDEX_OUT_OF_BOUNDS);
+        }
+        this.list.get(index).unmarkAsDone();
+        this.exportData();
+        return String.format("OK, I've marked this task as not done yet:%n%s", this.list.get(index).toString());
+    }
+
+    /**
+     * Deletes a task from the TaskList.
+     *
+     * @param num The number of the task to be deleted.
+     * @return String message
+     */
+    public String delete(int num) throws DukeIllegalArgumentException {
+        int index = num - 1;
+        if (index < 0 || index >= list.size()) {
+            throw new DukeIllegalArgumentException(ERROR_MESSAGE_INDEX_OUT_OF_BOUNDS);
+        }
+        Task task = list.remove(index);
+        this.exportData();
+        return String.format(DELETE_TASK_TEMPLATE, task, this.list.size(), taskOrTasks());
+    }
+
+    /**
+     * Prints tasks that contain the keyword in the description.
+     *
+     * @param keyword The keyword to search for.
+     * @return The String representation of the tasks that contain the keyword in the description.
+     */
+    public String find(String keyword) {
+        StringBuilder sb = new StringBuilder();
+        int count = 0;
+        for (int i = 0; i < list.size(); i++) {
+            Task task = list.get(i);
+            if (task.hasKeywordInDescription(keyword)) {
+                count++;
+                sb.append("\n").append(i + 1).append(".").append(task);
+            }
+        }
+        if (count == 0) {
+            return "There are no matching tasks in your list.";
+        } else {
+            String isOrAre = count == 1 ? "is" : "are";
+            String taskOrTasks = count == 1 ? "task" : "tasks";
+            return String.format("Here %s the %d matching %s in your list:%s", isOrAre, count, taskOrTasks, sb);
+        }
+    }
+
+    /**
+     * Returns the String representation of the TaskList.
+     *
+     * @return String representation of the TaskList.
+     */
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("Here are the tasks in your list:");
+        if (list.isEmpty()) {
+            sb.append("\nYou have no tasks in your list.");
+        } else {
+            for (int i = 0; i < list.size(); i++) {
+                sb.append("\n").append((i + 1)).append(".").append(list.get(i));
```

Comment:
Would it be possible to minimise the chaining of the appends?

Reply:
Okay, will take note in the next commit to increase readability. Thank you.